From: Len Brown <len.brown@intel.com>
Date: Mon, 13 May 2019 13:58:47 -0400
Subject: cpu/topology: Export die_id
Git-repo: git://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git
Git-commit: 0e344d8c709fe01d882fc0fb5452bedfe5eba67a
Patch-mainline: Queued in subsystem maintainer repository
References: jsc#SLE-5454

Export die_id in cpu topology, for the benefit of hardware that has
multiple-die/package.

Signed-off-by: Len Brown <len.brown@intel.com>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Reviewed-by: Ingo Molnar <mingo@kernel.org>
Acked-by: Peter Zijlstra (Intel) <peterz@infradead.org>
Cc: linux-doc@vger.kernel.org
Link: https://lkml.kernel.org/r/e7d1caaf4fbd24ee40db6d557ab28d7d83298900.1557769318.git.len.brown@intel.com
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 Documentation/cputopology.txt |   15 ++++++++++++---
 drivers/base/topology.c       |    4 ++++
 include/linux/topology.h      |    3 +++
 3 files changed, 19 insertions(+), 3 deletions(-)

--- a/Documentation/cputopology.txt
+++ b/Documentation/cputopology.txt
@@ -8,6 +8,12 @@ to /proc/cpuinfo output of some architec
 	socket number, but the actual value is architecture and platform
 	dependent.
 
+die_id:
+
+	the CPU die ID of cpuX. Typically it is the hardware platform's
+	identifier (rather than the kernel's).  The actual value is
+	architecture and platform dependent.
+
 2) /sys/devices/system/cpu/cpuX/topology/core_id:
 
 	the CPU core ID of cpuX. Typically it is the hardware platform's
@@ -77,6 +83,7 @@ they reflect the cpu and cache hierarchy
 For an architecture to support this feature, it must define some of
 these macros in include/asm-XXX/topology.h:
 #define topology_physical_package_id(cpu)
+#define topology_die_id(cpu)
 #define topology_core_id(cpu)
 #define topology_book_id(cpu)
 #define topology_drawer_id(cpu)
@@ -94,9 +101,11 @@ To be consistent on all architectures, i
 provides default definitions for any of the above macros that are
 not defined by include/asm-XXX/topology.h:
 1) physical_package_id: -1
-2) core_id: 0
-3) sibling_cpumask: just the given CPU
-4) core_cpumask: just the given CPU
+2) die_id: -1
+3) core_id: 0
+4) sibling_cpumask: just the given CPU
+5) core_cpumask: just the given CPU
+6) die_cpumask: just the given CPU
 
 For architectures that don't support books (CONFIG_SCHED_BOOK) there are no
 default definitions for topology_book_id() and topology_book_cpumask().
--- a/drivers/base/topology.c
+++ b/drivers/base/topology.c
@@ -58,6 +58,9 @@ static ssize_t name##_list_show(struct d
 define_id_show_func(physical_package_id);
 static DEVICE_ATTR_RO(physical_package_id);
 
+define_id_show_func(die_id);
+static DEVICE_ATTR_RO(die_id);
+
 define_id_show_func(core_id);
 static DEVICE_ATTR_RO(core_id);
 
@@ -87,6 +90,7 @@ static DEVICE_ATTR_RO(drawer_siblings_li
 
 static struct attribute *default_attrs[] = {
 	&dev_attr_physical_package_id.attr,
+	&dev_attr_die_id.attr,
 	&dev_attr_core_id.attr,
 	&dev_attr_thread_siblings.attr,
 	&dev_attr_thread_siblings_list.attr,
--- a/include/linux/topology.h
+++ b/include/linux/topology.h
@@ -184,6 +184,9 @@ static inline int cpu_to_mem(int cpu)
 #ifndef topology_physical_package_id
 #define topology_physical_package_id(cpu)	((void)(cpu), -1)
 #endif
+#ifndef topology_die_id
+#define topology_die_id(cpu)			((void)(cpu), -1)
+#endif
 #ifndef topology_core_id
 #define topology_core_id(cpu)			((void)(cpu), 0)
 #endif
